File: src/balancedBinaryTree.cpp
#include <iostream>
using namespace std;

struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode() : val(0), left(nullptr), right(nullptr) {}
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
 
class Solution {
public:
    bool isBalanced(TreeNode* root) 
    {
        if (!root) 
        {
            return true;
        }

        return CheckHeight(root) != -1;
    }

    int CheckHeight(TreeNode* root) 
    {
        if (!root)
        {
            return 0;
        }
        int leftHeight = CheckHeight(root->left);
        int rightHeight = CheckHeight(root->right);

        if (abs(leftHeight - rightHeight) > 1 || leftHeight == -1 || rightHeight == -1)
        {
            return -1;
        }

        return 1 + max(leftHeight, rightHeight);
    }
};

File: src/LeetCodePY/balancedBinaryTree.py
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        if not root:
            return True

        return self.CheckHeight(root) != -1

    def CheckHeight(self, root: Optional[TreeNode]) -> int:

        if not root:
            return 0

        left = self.CheckHeight(root.left)
        right = self.CheckHeight(root.right)

        if abs(left - right) > 1 or left == -1 or right == -1:
            return -1

        return max(left, right) + 1